File: csharp/leet_code/1971.cs
/// <summary>
/// This class provides a solution for the problem of checking if there is a valid path in a graph.
/// A path in a graph is a sequence of vertices where each pair of adjacent vertices is connected by an edge.
/// </summary>
public class Solution
{
    /// <summary>
    /// This method checks if there is a valid path between the source and destination vertices in the graph.
    /// </summary>
    /// <param name="n">The number of vertices in the graph.</param>
    /// <param name="edges">An array of edge pairs where each pair is an edge connecting two vertices in the graph.</param>
    /// <param name="source">The source vertex.</param>
    /// <param name="destination">The destination vertex.</param>
    /// <returns>True if there is a valid path from source to destination, false otherwise.</returns>
    public bool ValidPath(int n, int[][] edges, int source, int destination)
    {
        int[] parent = new int[n]; // An array to store the parent of each vertex in the graph.

        // Initialize each vertex to be its own parent.
        for (int i = 0; i < n; i++)
            parent[i] = i;

        // For each edge in the graph, find the parents of the two vertices and if they are not the same, union them.
        foreach (var edge in edges)
        {
            int parent1 = FindParent(parent, edge[0]);
            int parent2 = FindParent(parent, edge[1]);
            if (parent1 != parent2)
                parent[parent2] = parent1;
        }

        // Check if the source and destination vertices are in the same connected component of the graph.
        return FindParent(parent, source) == FindParent(parent, destination);
    }

    /// <summary>
    /// This method finds the parent of a vertex in the graph using path compression technique.
    /// </summary>
    /// <param name="parent">An array where the index is a vertex and the value at the index is the parent of the vertex.</param>
    /// <param name="node">The vertex whose parent is to be found.</param>
    /// <returns>The parent of the vertex.</returns>
    private int FindParent(int[] parent, int node)
    {
        // If the parent of the vertex is not itself, recursively find the parent.
        if (parent[node] != node)
            parent[node] = FindParent(parent, parent[node]);

        return parent[node]; // Return the parent of the vertex.
    }
}

File: go/leet_code/1971.go
// Package leet_code provides solutions for LeetCode problems.
package leet_code

// disjointSet represents a disjoint-set data structure, also known as a union-find data structure or merge-find set.
// It provides operations for adding new sets, merging sets (replacing them by their union), and finding a representative member of a set.
// The ranks array is used to optimize the union operation.
type disjointSet struct {
	roots []int // roots array represents the parent of each element in the set.
	ranks []int // ranks array stores the rank of each element.
}

// find method implements the find operation for the disjoint-set data structure.
// It returns the representative (root) of the set that element x belongs to.
// Path compression technique is used here to optimize subsequent find operations.
func (ds *disjointSet) find(x int) int {
	if ds.roots[x] == x {
		return x
	}
	ds.roots[x] = ds.find(ds.roots[x])
	return ds.roots[x]
}

// union method implements the union (or merge) operation for the disjoint-set data structure.
// It replaces the set containing element x and the set containing element y with their union.
// Union by rank is used here to optimize the union operation.
func (ds *disjointSet) union(x, y int) {
	rootX := ds.find(x)
	rootY := ds.find(y)
	if rootX == rootY {
		return
	}
	rankX := ds.ranks[rootX]
	rankY := ds.ranks[rootY]
	if rankX > rankY {
		ds.roots[rootY] = rootX
		return
	}
	if rankX < rankY {
		ds.roots[rootX] = rootY
		return
	}
	ds.roots[rootX] = rootY
	ds.ranks[rootY] += 1
}

// isConnected method checks if two elements x and y are in the same set.
// It returns true if x and y are in the same set, otherwise false.
func (ds *disjointSet) isConnected(x, y int) bool {
	return ds.find(x) == ds.find(y)
}

// newDisjointSet function creates a new disjoint-set data structure with n elements.
// Initially, each element is in a separate set.
func newDisjointSet(n int) *disjointSet {
	roots := make([]int, n)
	ranks := make([]int, n)
	for i := range n {
		roots[i] = i
		ranks[i] = 1
	}
	newDisjointSet := disjointSet{roots, ranks}
	return &newDisjointSet
}

// validPath function checks if there is a valid path between source and destination in the given graph.
// The graph is represented by n nodes and a list of edges.
// It uses the disjoint-set data structure to check the connectivity between source and destination.
// It returns true if there is a valid path, otherwise false.
func validPath(n int, edges [][]int, source int, destination int) bool {
	ds := newDisjointSet(n)
	for _, edge := range edges {
		ds.union(edge[0], edge[1])
	}
	return ds.isConnected(source, destination)
}
